driver/src/snapp.rs

```diff
@@ -0,0 +1,409 @@
+                            .ok_or(SnappObjectiveError::TokenNotFound)?,
+                        *self
+                            .executed_buy_amounts
+                            .get(i)
+                            .ok_or(SnappObjectiveError::OrderBookMismatch)?,
+                        *self
+                            .executed_sell_amounts
+                            .get(i)
+                            .ok_or(SnappObjectiveError::OrderBookMismatch)?,
+                    )?,
+                )
+                .ok_or(SnappObjectiveError::TotalUtilityOverflow)?;
+        }
+
+        Ok(total_executed_utility)
+    }
+}
+
+/// Snapp specific extensions for `Order`. This extension trait provides the
+/// required functions for calculating the objective value used in the Snapp
+/// protocol which is needed for solution submission.
+pub trait SnappOrder {
+    /// Returns the executed utility of an order based on executed amounts
+    /// and closing prices.
+    ///
+    /// Note that utility is calculated with the following equasion:
+    /// `((exec_buy_amt * sell_amt - exec_sell_amt * buy_amt) * buy_price) /
+    /// sell_amt`
+    ///
+    /// # Returns
+    ///
+    /// `Ok(U256)` when the order executed utility is correctly calculated
+    /// and `Err` otherwise.
+    fn executed_utility(
+        &self,
+        buy_price: u128,
+        exec_buy_amt: u128,
+        exec_sell_amt: u128,
+    ) -> Result<U256, SnappObjectiveError>;
+}
+
+impl SnappOrder for Order {
+    fn executed_utility(
+        &self,
+        buy_price: u128,
+        exec_buy_amt: u128,
+        exec_sell_amt: u128,
+    ) -> Result<U256, SnappObjectiveError> {
+        let buy_price = u128_to_u256(buy_price);
+        let exec_buy_amt = u128_to_u256(exec_buy_amt);
+        let exec_sell_amt = u128_to_u256(exec_sell_amt);
+        let buy_amt = u128_to_u256(self.buy_amount);
+        let sell_amt = u128_to_u256(self.sell_amount);
+
+        // the utility is caculated in two parts, this is done to avoid overflows
+        let utility_with_error = exec_buy_amt
+            .checked_sub(
+                (exec_sell_amt * buy_amt)
+                    .checked_div(sell_amt)
+                    .ok_or(SnappObjectiveError::InvalidOrder)?,
+            )
+            .ok_or(SnappObjectiveError::LimitPriceNotRespected)?
+            * buy_price;
+        let utility_error = (((exec_sell_amt * buy_amt) % sell_amt) * buy_price) / sell_amt;
+
+        let utility = utility_with_error
+            .checked_sub(utility_error)
+            .ok_or(SnappObjectiveError::LimitPriceNotRespected)?;
+
+        Ok(utility)
+    }
+}
+
+/// Represents and error that can occur during the calculation of the Snapp
+/// objective value.
+#[derive(Clone, Copy, Debug, Eq, Error, PartialEq)]
+pub enum SnappObjectiveError {
+    /// A token referenced by an order was not found in the solution.
+    #[error("token not found")]
```

Comment:
Nice, we should probably use this for the other error cases as well. Does it allow to chain errors, so that it will be easy to debug where a certain error came from (assuming multiple callsites could throw the same error type)?

Reply:
It does. Although it does not do automatic boxing so your error types can get pretty big if you are not careful.

Reply:
I can look into it as a separate PR.